### PR TITLE
Add support for mad_hi and mul_hi.

### PR DESCRIPTION
**Pull request by porcuquine**
This PR adds the `mul_hi` and `mad_hi` primitives for the OpenCL backend. Emulation in the interpreter and C backend works as expected. The C Sharp and Python backends are not implemented here.

### CODE REVIEW

**Comment by athas** [comment]
Not bad!  And you even made the tests a lot better than the ones for the other primitives.  I have a few style remarks (and maybe CI will have a few more), but it looks good.

File: src/Futhark/Representation/Primitive.hs [hunk 1]
```diff
@@ -1134,3 +1184,39 @@ convOp s from to = text s <> text "_" <> ppr from <> text "_" <> ppr to
 prettySigned :: Bool -> PrimType -> String
 prettySigned True (IntType it) = 'u' : drop 1 (pretty it)
 prettySigned _ t = pretty t
+
+mul_hi8 :: IntValue -> IntValue -> Int8
+mul_hi8 a b =
+  let a' = intToWord64 a in
+    let b' = intToWord64 b in
+      fromIntegral (shiftR (a' * b') 8)
```
Comment (athas) [inline src/Futhark/Representation/Primitive.hs]: More idiomatic Haskell:

```haskell
  let a' = intToWord64 a
      b' = intToWord64 b
  in fromIntegral (shiftR (a' * b') 8)
```

Comment (porcuquine) [reply src/Futhark/Representation/Primitive.hs]: Thanks, fixed.

File: tests/primitive/mad_hi.fut [hunk 2]
```diff
@@ -0,0 +1,88 @@
+-- Test u8.mad_hi
+-- tags { no_csharp no_python}
```
Comment (athas) [inline tests/primitive/mad_hi.fut]: The `tags` field need to be after the `==` line.

Comment (porcuquine) [reply tests/primitive/mad_hi.fut]: Fixed and fixed some test failures introduced when previously fixing parsing errors.

**Pull request merged**